### PR TITLE
Integrate xdoctest - Rebased

### LINKED ISSUE #71105: [docs] Make examples in docs into auto-runnable smoke tests
### 📚 The doc issue

Original discussion: https://github.com/pytorch/pytorch/issues/70972#issuecomment-1007444643 cc @brianjo @mruberry @dagitses . This will prevent outdated code examples (think deprecated functions / renamed arg names etc)

### Suggest a potential alternative/fix

More advanced version of this would be also verifying printed tensor outputs in docs, but smoke tests should be the first step

### PR DESCRIPTION
**Pull request by Erotemic**
This is a new version of #15648 based on the latest master branch.

Unlike the previous PR where I fixed a lot of the doctests in addition to integrating xdoctest, I'm going to reduce the scope here. I'm simply going to integrate xdoctest, and then I'm going to mark all of the failing tests as "SKIP". This will let xdoctest run on the dashboards, provide some value, and still let the dashboards pass. I'll leave fixing the doctests themselves to another PR.

In my initial commit, I do the bare minimum to get something running with failing dashboards. The few tests that I marked as skip are causing segfaults. Running xdoctest results in 293 failed, 201 passed tests. The next commits will be to disable those tests. (unfortunately I don't have a tool that will insert the `#xdoctest: +SKIP` directive over every failing test, so I'm going to do this mostly manually.)

Fixes https://github.com/pytorch/pytorch/issues/71105

@ezyang

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
happy to merge this once it is passing tests

**Comment by Erotemic** [comment]
So, I did end up writing a script that inserts `# xdoctest: +SKIP` above every failing test. There are still a few issues I'm seeing on my end, which is due to `python setup.py develop` failing (does that not work to install pytorch in dev mode anymore?)

It also seems like there isn't a single source of truth for what packages are needed as test requirements? I'm not sure exactly where to add xdoctest as a dependency. I see a few places that looked like decent candidates, so I added them there, but it may be worth creating a top-level `requirements` folder, and then things like `requirements/runtime.txt`, `requirements/docs.txt`, `requirements/tests.txt` `requirements/tests-ios.txt` etc... can be specified all in a single place. Anyway, that's an aside. Any guidance on where the current best practices are to declare dependencies in this repo would be helpful.

**Comment by ezyang** [comment]
You're right, unfortunately there isn't a unified spot. requirements.txt is a good start, but what I personally do is I grep for expecttest and insert my new dep everywhere expecttest is being installed.

**Comment by ezyang** [comment]
As you can see in, e.g., https://docs-preview.pytorch.org/82797/generated/torch.set_default_dtype.html#torch.set_default_dtype the skip tokens show up in the final output documentation. Is there any way we can suppress them?

**Comment by ezyang** [comment]
If you don't have a working CUDA environment, I recommend disabling xdoctest in that environment, as you'll probably end up with a bunch of tests that you can't actually test

**Comment by Erotemic** [comment]
> If you don't have a working CUDA environment, I recommend disabling xdoctest in that environment, as you'll probably end up with a bunch of tests that you can't actually test

If there is an enviornment variable or command line flag or existence of a module that indicates if cuda is available, we can use an xdoctest directive to conditionally skip that test. Or if there is some other runtime check that is easy to perform (a file existing in PATH), I could add a directive to xdoctest that tests for it. Or its always possible to add a runtime check in the doctest itself: e.g.

```
import torch, pytest
... if torch.cuda.is_available() else pytest.skip() 
```

but something like: `# xdoctest: +REQUIRES(env:CUDAHOME)` or `# xdoctest: +REQUIRES(CUDA)` 
 (the latter of which does not exist yet) might be nicer. 

I do have CUDA locally, I'm just having an issue building for some reason. (getting: `/usr/include/c++/11/bits/std_function.h:530:146: error: parameter packs not expanded with ‘...’`)

> you can see in, e.g., https://docs-preview.pytorch.org/82797/generated/torch.set_default_dtype.html#torch.set_default_dtype the skip tokens show up in the final output documentation. Is there any way we can suppress them? 

Generally, fixing the test is the best way to do it. But it should also be possible to remove patterns via sphinx (or whatever is used to generate the docs) preprocessing. 

I'm not sure why that specific one failed, it might be an issue with how I inserted the skips. It looks like it would have passed. It works just fine when I make a mwe:

```
====== <exec> ======
* DOCTEST : /home/joncrall/local/mwe.py::foo:0, line 3 <- wrt source file
DOCTEST SOURCE
1 >>> # Python floats are interpreted as float32
2 >>> import torch
3 >>> torch.tensor([1.2, 3]).dtype
  torch.float32
DOCTEST STDOUT/STDERR
DOCTEST RESULT
* SUCCESS: /home/joncrall/local/mwe.py::foo:0
====== </exec> ======
============
```

so it likely failed for a different reason, and that skip can be removed.

**Comment by albanD** [comment]
I think these changes break a few things?
- See the search result
  - before: https://pytorch.org/docs/stable/search.html?q=make_dual&check_keywords=yes&area=default
  - after: https://docs-preview.pytorch.org/82797/search.html?q=make_dual&check_keywords=yes&area=default
  - The rendering of the code sample is quite bad
- The doctest skips are pretty intrusive and make the code samples wrong. For example https://docs-preview.pytorch.org/82797/generated/torch.autograd.forward_ad.dual_level.html?highlight=make_dual

Is that expected?

**Comment by Erotemic** [comment]
I don't think it is expected. I'm not sure why it broke the formatting so much in that first example, but I'll fix it. 

I think there must be some issue in the way I automatically inserted skips. I agree this is producing ugly code and that needs to get fixed. 

The problem with these examples is that they are incomplete to begin with (which is why it's so important that they run on CI). In the example you gave, the "your_fn", while I see what it's trying to represent, is never defined. So, the code can never run. If I was fixing this doctest I would rewrite it like:

```
>>> import torch
>>> from torch.autograd.forward_ad import *
>>> def your_fn(inp):
>>>     ... # DO SOMETHING INTERESTING HERE
>>>     return inp + 1
>>> x = torch.tensor([1])
>>> x_t = torch.tensor([1])
>>> with dual_level():
...   inp = make_dual(x, x_t)
...   # Do computations with inp
...   out = your_fn(inp)
...   _, grad = unpack_dual(out)
>>> grad is None
False
>>> # After exiting the level, the grad is deleted
>>> _, grad_after = unpack_dual(out)
>>> grad is None
True
```

Now, that's not what I propose to _actually_ change this one to. The  `from torch.autograd.forward_ad import *` is implicitly provided via xdoctest, and `import torch` is implicitly provided by the xdoctest global-exec which I set to always make that module available in every test. That seems to be the style torch is going for (although I do think its better when a use can copy/paste the code and they can verify it runs without having to decode implicit assumptions).

I can live with the implicit imports, but I think this example doesn't make any sense unless some dummy input is given for `your_fn`.

If you wanted to keep the code as is (without specifying `your_fn`), then the better way to write this manually would be:

```
>>> # xdoctest: +SKIP
>>> x = torch.tensor([1])
>>> x_t = torch.tensor([1])
>>> with dual_level():
...   inp = make_dual(x, x_t)
...   # Do computations with inp
...   out = your_fn(inp)
...   _, grad = unpack_dual(out)
>>> grad is None
False
>>> # After exiting the level, the grad is deleted
>>> _, grad_after = unpack_dual(out)
>>> grad is None
True
```

with the skip at the top. I can certainly look into sphinx filtering to always remove them, but I think when it's at the top it doesn't look that bad, and it does alert the user that this isn't an example they can run without doing any extra work. So I think there is some value in rendering the SKIP when the test actually needs to be skipped. Of course the better thing is to ensure tests work, so the directive never has to be there in the first place.

**Comment by Erotemic** [comment]
I've gone through and removed a lot of the oddly placed skips (e.g. when they got inserted between control flow). There are still a lot of autogenerated skips in the code, but they should be filtered out so they don't appear in the documentation now. In many cases I went through the skips and added a reason, so you can distinguish an autogenerated skip versus an explicit one where I identified a reason why we had to skip the test. A lot of these cases are undefined variables, where the example is really meant to be a stub, and it doesn't test anything. However, in these cases xdoctest still checks syntax, so I had to make a few tweaks.

I've also added a few other directives `xdoctest: +REQUIRES(module:torchdata)` and `xdoctest: +REQUIRES(--lapac)`, `xdoctest: +REQUIRES(env:CUBLAS_WORKSPACE_CONFIG)`, which will conditionally disable tests based on the modules in the pythonpath, the flags on the CLI, and environment variables. To get things passing on the CI we might need to make a few environment variables to indicate features that are available in each case, or we can just entirely skip the test. Again, in all cases the directives _should_ be filtered out of the sphinx docs (haven't verified this yet).

The remaining cases where skips exist are probably due to more undefined variables, but I also saw a few cases where the doctest was checking a non-deterministic value. I didn't do anything to fix those. I'd like to avoid as many fixes like that as possible just so the dashboards pass with a decent number of the doctests being run (and most of them being syntax checked). Fixing these would also be a good task for first-time contributors, so I'm punting on that aspect.

**Comment by Erotemic** [comment]
The filter seems to have worked to take the directives out of the main examples, but the search page itself still looks different. I'm not sure why, or if matters. It's strange because Sphinx should be getting the _exact_ same input now.

File: .circleci/docker/requirements-ci.txt
Comment (ezyang) [inline .circleci/docker/requirements-ci.txt:167]: you... commented this out? lol

**Comment by ezyang** [comment]
It looks like getting the CI to play ball will take some time; maybe we can avoid merge conflicts by merging a version of this PR that has the comment stripping and ignores, but WITHOUT turning on the xdoctests, and then merge the actual xdoctest runner afterwards.

**Comment by ezyang** [comment]
Search page being different appears to just be an artifact of the PR doc generation; here is another PR's doc page which also has similar artifacts: https://docs-preview.pytorch.org/82746/search.html?q=empty&check_keywords=yes&area=default

**Comment by Erotemic** [comment]
I think that's a good idea. I pushed a patch that comments out the places where xdoctest is enabled.

**Comment by ezyang** [comment]
test failures are still real.

also cc @vadimkantorov

**Comment by vadimkantorov** [comment]
this is very cool :) are all examples tested this way? if not - it would be cool to add some visual indication to all examples to show if it's being tested or not and if yes, which version of pytorch passed the test (commit id or just pytorch verison) - or maybe this can be put in tooltip / title html attribute (though not very discoverable)

**Comment by vadimkantorov** [comment]
Also, if examples are using torch.rand tensors and then if wanted to check their output (which is sometimes shown in the example)  can be brittle

**Comment by Erotemic** [comment]
@vadimkantorov Any test that was using a random tensor should have an `# xdoctest: +SKIP` directive above it somewhere, but beyond that, my last commit _should have_ prevented xdoctest from running at all! I'm 100% not sure why the current failures are happening.

Looking at one of the failures:

```
======================================================================
FAIL [0.008s]: test_docstring_examples (__main__.TestAssertClose)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/test_testing.py", line 811, in test_docstring_examples
    raise AssertionError(f"Doctest found {len(failures)} failures:\n\n" + "\n".join(failures))
AssertionError: Doctest found 1 failures:

**********************************************************************
File "/opt/conda/lib/python3.10/site-packages/torch/testing/_comparison.py", line 1307, in assert_close
Failed example:
    torch.testing.assert_close(actual, expected)
Exception raised:
    Traceback (most recent call last):
      File "/opt/conda/lib/python3.10/doctest.py", line 1346, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest assert_close[34]>", line 1, in <module>
        torch.testing.assert_close(actual, expected)
      File "/opt/conda/lib/python3.10/site-packages/torch/testing/_comparison.py", line 1344, in assert_close
        assert_equal(
      File "/opt/conda/lib/python3.10/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
        raise error_metas[0].to_error(msg)
    AssertionError: Scalars are not close!

    Absolute difference: nan (up to 1e-05 allowed)
    Relative difference: nan (up to 1.3e-06 allowed)
```

It seems like something else is trying to run the doctests. That is not an xdoctest output...

... And now I see it. The introduced `# xdoctest: +SKIP` is causing the regular doctest runner to fail to see its output. Moving those into the correct place will prevent that from happening..

**Comment by vadimkantorov** [comment]
do these tests also check for generated warnings? examples should not recommend code that produces warnings from outdated usages

**Comment by Erotemic** [comment]
The xdoctest runner (or pytest using the xdoctest plugin) will report all warnings at the end. It is the responsibility of the developer to fix those warnings as it would be with any other real code.

I've pushed up the patch that should fix issues where the builtin doctest module was still running some of the examples. I might have missed a case or two (there were a lot of them), but I expect a lot of the errors will be fixed.

I've also:

1. Fixed more doctests when it was obvious what the problem was (e.g. missing an ending `)`)
2. Re-enabled tests that were auto-marked as SKIP but seemed fine.
3. Re-enabled tests auto-skipped tests with non-determinism and instead marked them as IGNORE_WANT, so xdoctest wont check their output against the example. 
4. Added more reasons as to why a test is skipped in the directive.

I've also noticed that some of the doctests seem to modify global torch state, which means if I run them by themselves they pass, but if I run them together, something like the default precision might cause unexpected results to happen. I marked these tests as such, or introduced fuzzier matching to allow passing in these cases (not that the xdoctest runner is enabled in this PR anyway).

With these patches on my machine I now get: 

```
=== 316 passed, 229 skipped, 13 warnings in 70.16 seconds ===
```

@vadimkantorov the warnings emitted are:

```
=== Found 13 run-time warnings ===
--- Runtime Warning: 1 / 13 ---
example = <DocTest(torch.functional chain_matmul:0 ln 1539) at 0x7fcc62cf6d30>
    /home/joncrall/code/pytorch/torch/functional.py:1557: UserWarning: torch.chain_matmul is deprecated and will be removed in a future PyTorch release. Use torch.linalg.multi_dot instead, which accepts a list of two or more tensors rather than multiple parameters. (Triggered internally at /home/joncrall/code/pytorch/aten/src/ATen/native/LinearAlgebra.cpp:983.)
      return _VF.chain_matmul(matrices)  # type: ignore[attr-defined]
    
--- Runtime Warning: 2 / 13 ---
example = <DocTest(torch.hub list:0 ln 392) at 0x7fcbe3c61b20>
    /home/joncrall/.cache/torch/hub/pytorch_vision_main/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: 
      warn(f"Failed to load image Python extension: {e}")
    
--- Runtime Warning: 3 / 13 ---
example = <DocTest(torch.hub load:0 ln 523) at 0x7fcbe3c61610>
    /home/joncrall/.cache/torch/hub/pytorch_vision_main/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.
      warnings.warn(
    
    /home/joncrall/.cache/torch/hub/pytorch_vision_main/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=ResNet50_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet50_Weights.DEFAULT` to get the most up-to-date weights.
      warnings.warn(msg)
    
--- Runtime Warning: 4 / 13 ---
example = <DocTest(torch._tensor Tensor.refine_names:0 ln 1072) at 0x7fcc62bda1f0>
    /home/joncrall/code/pytorch/torch/_tensor.py:1090: UserWarning: Named tensors and all their associated APIs are an experimental feature and subject to change. Please do not use them for anything important until they are released as stable. (Triggered internally at /home/joncrall/code/pytorch/c10/core/TensorImpl.h:1525.)
      return super(Tensor, self).refine_names(names)
    
--- Runtime Warning: 5 / 13 ---
example = <DocTest(torch.nn.functional nll_loss:0 ln 2675) at 0x7fcbe3d3a6d0>
    <doctest:/home/joncrall/code/pytorch/test/../torch/nn/functional.py::nll_loss:0>:5: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.
    
--- Runtime Warning: 6 / 13 ---
example = <DocTest(torch.nn.quantized.dynamic.modules.conv Conv1d:0 ln 33) at 0x7fcbe3d09460>
    /home/joncrall/code/pytorch/torch/nn/quantized/dynamic/modules/conv.py:58: UserWarning: The current implementation of the DynamicQuantizedConv1d module has poor numerical accuracy and its use is not recommended
      warnings.warn(
    
--- Runtime Warning: 7 / 13 ---
example = <DocTest(torch.nn.quantized.dynamic.modules.conv Conv2d:0 ln 104) at 0x7fcbe3d098b0>
    /home/joncrall/code/pytorch/torch/nn/quantized/dynamic/modules/conv.py:123: UserWarning: The current implementation of the DynamicQuantizedConv2d module has poor numerical accuracy and its use is not recommended
      warnings.warn(
    
--- Runtime Warning: 8 / 13 ---
example = <DocTest(torch.nn.quantized.dynamic.modules.conv Conv3d:0 ln 169) at 0x7fcbe70fb670>
    /home/joncrall/code/pytorch/torch/nn/quantized/dynamic/modules/conv.py:188: UserWarning: The current implementation of the DynamicQuantizedConv3d module has poor numerical accuracy and its use is not recommended
      warnings.warn(
    
--- Runtime Warning: 9 / 13 ---
example = <DocTest(torch.nn.modules.activation Softmin:0 ln 1305) at 0x7fcbe3c57040>
    <doctest:/home/joncrall/code/pytorch/test/../torch/nn/modules/activation.py::Softmin:0>:3: UserWarning: Implicit dimension choice for softmin has been deprecated. Change the call to include dim=X as an argument.
    
--- Runtime Warning: 10 / 13 ---
example = <DocTest(torch.nn.modules.activation LogSoftmax:0 ln 1432) at 0x7fcbe3c57430>
    <doctest:/home/joncrall/code/pytorch/test/../torch/nn/modules/activation.py::LogSoftmax:0>:3: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.
    
--- Runtime Warning: 11 / 13 ---
example = <DocTest(torch.nn.modules.loss KLDivLoss:0 ln 451) at 0x7fcbe3c55100>
    <doctest:/home/joncrall/code/pytorch/test/../torch/nn/modules/loss.py::KLDivLoss:0>:3: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.
    
    <doctest:/home/joncrall/code/pytorch/test/../torch/nn/modules/loss.py::KLDivLoss:0>:5: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
    
    <doctest:/home/joncrall/code/pytorch/test/../torch/nn/modules/loss.py::KLDivLoss:0>:2: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.
    
--- Runtime Warning: 12 / 13 ---
example = <DocTest(torch.nn.modules.module Module.to:0 ln 923) at 0x7fcbe3d248b0>
    /home/joncrall/code/pytorch/torch/nn/modules/module.py:970: UserWarning: Complex modules are a new feature under active development whose design may change, and some modules might not work as expected when using complex tensors as parameters or buffers. Please file an issue at https://github.com/pytorch/pytorch/issues/new?template=bug-report.yml if a complex module does not work as expected.
      warnings.warn(
    
--- Runtime Warning: 13 / 13 ---
example = <DocTest(torch.autograd.anomaly_mode detect_anomaly:0 ln 23) at 0x7fcbe3d014c0>
    <doctest:/home/joncrall/code/pytorch/test/../torch/autograd/anomaly_mode.py::detect_anomaly:0>:1: UserWarning: Anomaly Detection has been enabled. This mode will increase the runtime and should only be enabled for debugging.
```

**Comment by vadimkantorov** [comment]
I would vote for fixing warnings if it's fast, and if not: publish somewhere a list of warning-throwing examples to fix all the warnings later

**Comment by vadimkantorov** [comment]
About global state: unfortunately, this is somewhat common in pytorch :( so probably independent Python instances would be needed to enforce re-init of pytorch at least for these kinds of tests

**Comment by Erotemic** [comment]
I've done a bunch of work on fixing the doctests in general, I don't think I have much more bandwidth to fix the warnings as well (plus they shouldn't even be triggered in this PR given that the xdoctest runner wont be enabled). So maybe we should defer that conversation until this branch is merged and the "enable xdoctest runner" PR is started.

For the global state, is there any way to reset it to the defaults? I think the best way to handle this is that the tests that change it should clean up after themselves. (It would also be valuable in those examples to show the user how to reset the state).

**Comment by ezyang** [comment]
> For the global state, is there any way to reset it to the defaults? I think the best way to handle this is that the tests that change it should clean up after themselves. (It would also be valuable in those examples to show the user how to reset the state).

This should be possible in most of the cases, I think.

**Comment by Erotemic** [comment]
@ezyang Dashboard is green

**Review by ezyang — APPROVED** [review]

**Comment by Erotemic** [comment]
I suppose the labels [topic: documentation](https://github.com/pytorch/pytorch/labels/topic%3A%20documentation) and / or [topic: not user facing](https://github.com/pytorch/pytorch/labels/topic%3A%20not%20user%20facing) should be applied. Not sure how to do that.

**Comment by kit1980** [comment]
> I suppose the labels [topic: documentation](https://github.com/pytorch/pytorch/labels/topic%3A%20documentation) and / or [topic: not user facing](https://github.com/pytorch/pytorch/labels/topic%3A%20not%20user%20facing) should be applied. Not sure how to do that.

You can use pytorchbot for this, see https://github.com/pytorch/pytorch/wiki/Bot-commands#labeling

**Pull request closed without merging**
